man/informant.1
.TH INFORMANT 1 "02 SEPTEMBER 2024" informant-v0.6.0 "Informant Manual"
.SH NAME
informant \- An Arch Linux News reader and pacman hook.

.SH SYNOPSIS
.I informant
[options] check

.I informant
[options] list [--reverse --unread]

.I informant
[options] read [<item> | --all]

.SH DESCRIPTION
.I informant
has three modes of operation: check, list and read.

.SH COMMANDS

.TP
.B check
Check for unread Arch Linux News items and exit with status code equal to the
number of unread items.

.TP
.B list
Print out the headlines of the most recent news items (regardless of whether or
not they have been read). list also can take two options of its own:
.RS
.TP
.B \-\-reverse
Print the headlines oldest to newest (instead of newest to oldest).
.TP
.B \-\-unread
Print the headlines of only unread items.
.RE

.TP
.BR read " " [<item> | --all]
Print the given news item and mark it as read.
.B <item>
can be given as an index (printed by the
.B list
command without the
.B \-\-unread
or
.B \-\-reverse
options) or as a string matching part of the headline of a news item.
.B read
will exit with status code 255 if it does not have permission to write to the
save file.

When run without
.BR <item>
or
.B \-\-all
read will run interactively printing all unread news items from oldest to newest
prompting the user if they want to continue reading.
.RS
.TP
.B \-\-all
Mark all the most recent news items as read without printing them.
.RE

.SH OPTIONS
These are the global options which can be applied to any of the subcommands.

.TP
.BR \-c " " <cfile> ", " \-\-config=<cfile>
Read configuration from <file>. Configuration and syntax can be found below in
CONFIGURATION.

.TP
.BR \-d ", " \-\-debug
Print the debug messages and don't make changes to the save file.

.TP
.BR \-r ", " \-\-raw
When printing items do not replace any markup, items will be printed exactly as
returned by the RSS feed.

.TP
.BR \-f " " <file> ", " \-\-file=<file>
Use <file> as the save location for marking items as read. This is a binary
file.

.B NOTE
Changing the file will not change the file read by the pacman hook because the
pacman hook runs as root, if you want to change that you can do so manually by
editing /usr/share/libalpm/hooks/00-informant.hook

.TP
.BR \-p " " <pager> ", " \-\-pager=<pager>
Use <pager> to display news items. This option overrides the
.B $INFORMANT_PAGER
environment variable.

.TP
.BR \-\-no\-cache
Do not consider the cache (/var/cache/informant/) when fetching feeds (this
could likely result in slower feed fetching but more up to date results). Using
this may be helpful to verify issues running informant due to the cache.

.TP
.BR \-\-clear-cache
Empty the cache directory (/var/cache/informant/) before fetching the feed(s).

.TP
.BR \-\-clear-readlist
Empty the save file of read news items (/var/lib/informant.dat by default), this
will mean that any previously read news items returned in the feed will need to
be read again. The news feed only returns a set amount of files and over time
there may end up being items in the save file that are no longer returned when
fetching the feed, this option can be used to cleanup the save file.

By clearing the save file and then marking everything as read you effectively
remove all the old items from the save file:
.EX
    informant --clear-readlist read --all
.EE

.TP
.BR \-h ", " \-\-help
Print the help and exit.

.TP
.BR \-V ", " \-\-version
Print the version information and exit.

.SH ENVIRONMENT

.TP
.BR INFORMANT_PAGER
If
.B $INFORMANT_PAGER
is set it's value will be used as the name of the program used to display news
items.

.SH CONFIGURATION
Informant can be configured to check multiple feeds instead of just the Arch
Linux News feed (whether having it do so is actually useful or not is left up to
the user). Informant will check for an
.B informantrc.json
file in a few places. It will check in this order:
.RS
.PP
\- CLI provided option (cfile)

.PP
\- $HOME/.informantrc.json

.PP
\- $XDG_CONFIG_HOME/informantrc.json

.PP
\- /etc/informantrc.json

.PP
\- For each directory $d in $XDG_CONFIG_DIRS it will look for $d/informantrc.json
.RE

.PP
.B NOTE
IF you want the configuration file to be used in the pacman hook make sure to
use a save location that will be accessible to informant when running as
root/sudo.

.SS Configuration Syntax
The syntax of informantrc.json is a JSON object containig on key:
.B feeds
that is a list describing the feeds you want informant to check. Each feed is
represented as a JSON object with the following keys:
.RS

.TP
.BR name (optional)
used to show which feed each news item is from

.TP
.BR url (required)
the feed URL

.TP
.BR title-key (optional)
defaults to
.IR title ,
the key used to reference the news item title in the feed

.TP
.BR body-key (optional)
defaults to
.IR summary ,
the key used to reference the news item body in the feed

.TP
.BR timestamp-key (optional)
defaults to
.IR published ,
the key used to reference the news item date in the feed
.RE

.PP
For example, the below informantrc.json will configure informant to check the
Arch Linux News feed as well as the Arch Linux 32 News Feed:
.EX
{
    "feeds": [
        {
            "name": "Arch News",
            "url": "https://archlinux.org/feeds/news",
            "title-key": "title",
            "body-key": "summary",
            "timestamp-key": "published"
        },
        {
            "name": "Arch32 News",
            "url": "https://bbs.archlinux32.org/extern.php?action=feed&fid=12&type=atom",
            "title-key": "title",
            "body-key": "summary",
            "timestamp-key": "updated"
        }
    ]
}
.EE

.SH AUTHOR
Bradford Smith <\fIhttps://github.com/bradford-smith94\fR>

.SH LICENSE
MIT.

.SH BUGS
Please report bugs at:
<\fIhttps://github.com/bradford-smith94/informant/issues\fR>.

.SH "SEE ALSO"
.BR alpm-hooks (5)


informant/informant.py
#!/usr/bin/python3
"""
informant - an Arch Linux News reader designed to also be used as a pacman hook

Usage:
    informant [options] check
    informant [options] list [--reverse --unread]
    informant [options] read [<item> | --all]

Commands:
    check - Check for unread news items, will exit with a positive return code
            with the number of unread items. If there is only one unread item it
            will also print in like 'read' would and mark it as read, else it
            will print the number of unread titles.

    list -  Print the most recent news items, regardless of read status. If
            '--reverse' is provided items are printed oldest to newest. If
            '--unread' is provided only the unread items are returned.

    read  - Read the specified news item, <item> can be either an index or a
            full title. This will also save the item as 'read' so that future
            calls to 'check' will no longer display it. If no <item> is given,
            will begin looping through all unread items, printing each one and
            marking them as read with a prompt to continue. Passing the --all
            flag will mark all items as read without printing.

Options:
    -c <cfile>, --config=<cfile>    Use <cfile> as the config file
    -d, --debug                     Print the debug messages and don't make
                                    changes to the save file
    -r, --raw                       When printing items do not replace any
                                    markup
    -f <file>, --file=<file>        Use <file> as the save location for read
                                    items
    -p <pager>, --pager=<pager>     Use <pager> as the pager to display the
                                    news items
    --no-cache                      Do not use cache
    --clear-cache                   Empty the cache before fetching feed(s)
    --clear-readlist                Empty the saved readlist
    -h, --help                      Show this help and exit
    -V,--version                    Show version and exit

"""

# builtins
import sys

# external
import docopt

# local
from informant.config import InformantConfig
from informant.feed import Feed
import informant.file as fs
import informant.ui as ui

__version__ = '0.6.0'

# commands
CHECK_CMD = 'check'
LIST_CMD = 'list'
READ_CMD = 'read'

# 'list' options
REV_OPT = '--reverse'
UNREAD_OPT = '--unread'

# 'read' options and args
ITEM_ARG = '<item>'
READALL_OPT = '--all'

def check_cmd(feed):
    """ Run the check command. Check if there are any news items that are
    unread. If there is only one unread item, print it out and mark it as read.
    Also, exit the program with return code matching the unread count. """
    running_from_pacman = ui.running_from_pacman()
    unread = 0
    unread_items = []
    for entry in feed:
        if not entry.has_been_read():
            unread += 1
            unread_items.append(entry)
    if unread == 1:
        if running_from_pacman:
            ui.pacman_msg('Stopping upgrade to print news')
        ui.pretty_print_item(unread_items[0])
        unread_items[0].mark_as_read()
        fs.save_datfile()
        if running_from_pacman:
            ui.pacman_msg('You can re-run your pacman command to complete the upgrade')
    elif unread > 1:
        print('There are {:d} unread news items! Use informant to read \
them.'.format(unread))
        if running_from_pacman:
            ui.pacman_msg('Run `informant read` before re-running your pacman command')
    else:
        print('There are no unread news items')
    sys.exit(unread)

def list_cmd(feed):
    """ Run the list command. Print out a list of recent news item titles. """
    argv = InformantConfig().get_argv()
    if argv.get(REV_OPT):
        feed_list = reversed(feed)
    else:
        feed_list = feed
    index = 0
    for entry in feed_list:
        if not argv.get(UNREAD_OPT) \
        or (argv.get(UNREAD_OPT) and not entry.has_been_read()):
            print(ui.format_list_item(entry, index))
            index += 1

def read_cmd(feed):
    """ Run the read command. Print news items and mark them as read. """
    argv = InformantConfig().get_argv()
    if argv.get(READALL_OPT):
        for entry in feed:
            entry.mark_as_read()
    else:
        if argv[ITEM_ARG]:
            try:
                item = int(argv[ITEM_ARG])
                entry = feed[item]
            except ValueError:
                for entry in feed:
                    if entry.title == item:
                        break
                #NOTE: this will read the oldest unread item if no matches are found
            ui.pretty_print_item(entry)
            entry.mark_as_read()
        else:
            unread_entries = list()
            for entry in feed:
                if not entry.has_been_read():
                    unread_entries.insert(0, entry)
            for entry in unread_entries:
                ui.pretty_print_item(entry)
                entry.mark_as_read()
                if entry is not unread_entries[-1]:
                    read_next = ui.prompt_yes_no('Read next item?', 'yes')
                    if read_next in ('n', 'no'):
                        break
                else:
                    print('No more unread items')
    fs.save_datfile()

def main():
    """ The main function.
    Check given arguments get feed and run given command. """
    argv = docopt.docopt(__doc__, version='informant v{}'.format(__version__))
    InformantConfig().set_argv(argv)
    InformantConfig().debug_print = ui.debug_print
    InformantConfig().readlist = fs.read_datfile()
    config = InformantConfig().get_config()
    ui.debug_print('cli args: {}'.format(argv))

    if 'feeds' in config:
        feed = []
        for config_feed in config['feeds']:
            feed += Feed(config_feed).entries
    else:
        feed = Feed().entries

    if not feed:
        ui.warn_print('no news feed items, informant is performing no action')
        sys.exit()

    feed = sorted(feed, key=lambda k: k.timestamp, reverse=True)

    if argv.get(CHECK_CMD):
        check_cmd(feed)
    elif argv.get(LIST_CMD):
        list_cmd(feed)
    elif argv.get(READ_CMD):
        read_cmd(feed)
    sys.exit()

if __name__ == '__main__':
    main()
